### ISSUE TITLE
Presets, including a Vim Preset

### ISSUE THREAD

**Issue by brillout** [johtela/vscode-modaledit#7]
Hi @johtela !

I'm a long-time Vim user and I just started to use VS Code because I'm starting to use TypeScript.

What you are building looks really nice, it seems to me that you're onto something.

What I like:
 - 100% VS Code experience when in insert mode (actually, I'm not entirely sure it's the case but that's what I've interpreted so far.)
 - Fully flexible. E.g. ModalEdit has zero effect when there are no keybindings defined.
 - You want to go beyond what vim does. Vim was only the beginning and modal editing has a bright future.
 - You seem to care about what you are building and looks like you're in this for the long run.

In a nutshell, ModalEdit simply adds a modal mode and gets out of the way otherwise. That's exactly what I'm looking for; I want to fully embrace vscode while having the modal efficiency that I have grown to love with vim.

Am I right at depicting your vision?

Anyways, back to the topic.

My fingers are hard-coded to a bunch of Vim shortcuts; I'd like to start using ModalEdit with a Vim Preset that presets the most common vim keybindings. Over time I'd like to diverge from Vim keybinding to my own.

What do you think of adding presets? For people like me coming directly from vim, a vim preset would be super convenient to get started quickly. In the future it would be nice to have a multitude of presets that can be combined. For example:
 - day 0: I start with the vim preset only.
 - day 10: I add the "Quick Snippets Preset" that adds your Quick Snippets feature. (Looks like a very neat feature but I'm overwhelmed with all your lovely features right now; I'd like to get started quickly without any of your advanced features but progressively learn about them.)
 - day 20: I start my own "Romuald Preset"
 - Etc.

At day 20 I have 3 presets and the last "Romuald Preset" overrides all previous presets.

What do you think?

**Comment by Delapouite** [comment]
Your approach definitely makes sense.

But it should be packaged with some disclaimer, saying that the vim preset only covers the basics and do not intend to be a perfect emulation. 
Because unfortunately people can be very nitpicking about vim implementations in editors and would complain that this particular feature is missing or the behavior is not 100% the same than in the "one true vim". 
The end goal should be that it does not turn into a burden for the maintainer of the preset.

**Comment by brillout** [comment]
@Delapouite Good point & I agree. Personally, I don't want a perfect clone but I'd love something better! Hehe :)

**Comment by brillout** [comment]
>  I'd love something better

Which actually would already be the case with 100% VS Code insert mode + basic vim commands in normal mode

**Comment by johtela** [comment]
Hi @brillout,

Thanks for the positive feedback. You are 100% right about the vision. The idea of ModalEdit is to enable Vim features without changing the core functionality of VS Code. You are also correct in assuming that insert mode works as vanilla VS code without the extension.

Regarding presets, the basic Vim commands are (kind of) already available [here](https://raw.githubusercontent.com/johtela/vscode-modaledit/master/.vscode/settings.json). This is the literate source file of the [tutorial page](https://johtela.github.io/vscode-modaledit/docs/.vscode/settings.html). These bindings do not cover all the possible Vim commands that you can emulate with ModalEdit, but they should give you an idea how to implement the missing features.

I can definitively see the need for having sensible presets. As you start with a blank slate, there is quite a lot of work involved in getting you up and running. But I would not want to impose those presets automatically to all users, they should be something you opt-in to.

The thing is, although ModalEdit is not meant to be a Vim clone, people finding it useful are mostly Vim users who have migrated to VS Code, and are dissatisfied with existing emulations. That was my motivation, anyway 😉 So, it makes sense to cater for these users. In fact, I am currently thinking of a simple way to add Vim's "repeat last change" command to ModalEdit. But as @Delapouite pointed out Vim users tend to be very picky that things work exactly like in Vim, and providing a full Vim emulation out-the-box is not a goal of ModalEdit. Presets should be just a starting point or example to teach you how to define your own bindings.

If you have any ideas or suggestions how these presets could be distributed or deployed, let me know. The simplest option is to expand the  bindings in the tutorial and make them available in GitHub (without the comments). I could also include some additional examples that contain bindings that I have found useful myself (my own bindings are available [here](https://gist.github.com/johtela/b63232747fdd465748fedb9ca6422c84)).

**Comment by brillout** [comment]
I agree with everything you say :-).

How about the following?

A `modaledit` directory next to `settings.json` with following files:
 - `modaledit/modaledit.js` that contains all ModalEdit settings
 - `modaledit/vim.js` a vim preset
 - `modaledit/johtela.js` your own preset

~~~js
// mdaledit/modaledit.js

module.exports = {
  presets: [
    './vim.js',
    './johtela.js',
  ],
};
~~~

~~~js
module.exports = {
   keybindings: {
     // vim preset copy pasted from github.com/johtela/vscode-modaledit
  }
};
~~~

To keep things simple at the beginning, the user would copy paste the Vim preset from a file in this repository.

I purposely made these file JavaScript files and not JSON files, see #9.

**Comment by johtela** [comment]
Thanks for the suggestions. Since doing a semi-complete implementation of Vim keybindings is quite a large undertaking, I can't promise it will be ready soon. I will try to include them to the next minor release.

**Comment by brillout** [comment]
That would be absolutely amazing.

It takes time to change hard-wired typing habits and a Vim preset would help a lot.

> semi-complete

Personally, I'm totally fine if only the most useful Vim keybindings are supported. Anything I can do with vscode, I'd rather do with vscode directly. For example I'd rather use vscode's multi cursor feature than something like vim's macro.

And actually, to me, what you did in https://raw.githubusercontent.com/johtela/vscode-modaledit/master/.vscode/settings.json is almost it. I'm currently using this "preset" and I'm already fairly happy. The only thing I'm missing are the "ci", "di", and "yi" features.


Beyond that, I'm looking forward to "progressively migrate" towards your modal editing world :-).

**Comment by johtela** [comment]
Vim presets are included in version 2.0. See the [documentation](https://johtela.github.io/vscode-modaledit/docs/presets/vim.html) for details.
